src/device_trezor/trezor/transport.hpp

```diff
@@ -0,0 +1,331 @@
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#ifndef MONERO_TRANSPORT_H
+#define MONERO_TRANSPORT_H
+
+
+#include <boost/asio.hpp>
+#include <boost/asio/deadline_timer.hpp>
+#include <boost/array.hpp>
+#include <boost/utility/string_ref.hpp>
+
+#include <typeinfo>
+#include <type_traits>
+#include "net/http_client.h"
+
+#include "rapidjson/document.h"
+#include "rapidjson/writer.h"
+#include "rapidjson/stringbuffer.h"
+
+#include "exceptions.hpp"
+#include "trezor_defs.hpp"
+#include "messages_map.hpp"
+
+#include "messages/messages.pb.h"
+#include "messages/messages-common.pb.h"
+#include "messages/messages-management.pb.h"
+#include "messages/messages-monero.pb.h"
+
+namespace hw {
+namespace trezor {
+
+  using json = rapidjson::Document;
+  using json_val = rapidjson::Value;
+  namespace http = epee::net_utils::http;
+
+  const std::string DEFAULT_BRIDGE = "127.0.0.1:21325";
+
+  // Base HTTP comm serialization.
+  bool t_serialize(const std::string & in, std::string & out);
+  bool t_serialize(const json_val & in, std::string & out);
+  std::string t_serialize(const json_val & in);
+
+  bool t_deserialize(const std::string & in, std::string & out);
+  bool t_deserialize(const std::string & in, json & out);
+
+  // Flexible json serialization. HTTP client tailored for bridge API
+  template<class t_req, class t_res, class t_transport>
+  bool invoke_bridge_http(const boost::string_ref uri, const t_req & out_struct, t_res & result_struct, t_transport& transport, const boost::string_ref method = "POST", std::chrono::milliseconds timeout = std::chrono::seconds(180))
+  {
+    std::string req_param;
+    t_serialize(out_struct, req_param);
+
+    http::fields_list additional_params;
+    additional_params.push_back(std::make_pair("Origin","https://monero.trezor.io"));
```

Comment:
Intended ? Or should this be the hostname passed as param (so 127.0.0.1) ?

Reply:
This is a particular requirement of the Trezor Bridge. It requires the Origin of the form *.trezor.io. 
Is that right, @prusnak ?

Reply:
If so, does it do anything with it ? Does it try to connect ? If it does, does it die if internet access is off ? If the trezor.io domain goes ?

Reply:
@prusnak could explain more, but I guess it is just some kind of identificator of the client.

Reply:
@moneromooo-monero this is just the "Origin" HTTP header. trezor bridge discards any requests that do not come from trezor.io domain. If you are developing an app that is not a webapp, you can spoof origin, so it does work.